proof-systems/src/darlin/tests/final_darlin.rs

```diff
@@ -0,0 +1,358 @@
+//! A test circuit which, besides processing incremental data according to 
+//! a simple quadratic relation, allocates a given instance of `FinalDarlinDeferredData`,
+//! and wires it to the outside via system inputs.
+use algebra::{AffineCurve, ToConstraintField, UniformRand};
+use r1cs_core::{ConstraintSynthesizer, ConstraintSystem, SynthesisError};
+use crate::darlin::{
+    pcd::{
+        PCD, PCDParameters, PCDCircuit,
+        final_darlin::FinalDarlinPCD,
+        error::PCDError,
+    },
+    accumulators::ItemAccumulator,
+    data_structures::FinalDarlinDeferredData,
+    FinalDarlinProverKey, FinalDarlinVerifierKey, FinalDarlin,
+};
+use poly_commit::{
+    ipa_pc::{InnerProductArgPC, CommitterKey, UniversalParams},
+    Error as PCError
+};
+use rand::{ Rng, RngCore };
+use digest::Digest;
+use r1cs_std::{
+    alloc::AllocGadget,
+    fields::fp::FpGadget,
+    eq::EqGadget,
+};
+use std::ops::MulAssign;
+
+// Dummy Acc used for testing
+pub struct TestAcc {}
+
+impl ItemAccumulator for TestAcc {
+    type AccumulatorProverKey = ();
+    type AccumulatorVerifierKey = ();
+    type AccumulationProof = ();
+    type Item = ();
+
+    fn check_items<R: RngCore>(
+        _vk: &Self::AccumulatorVerifierKey,
+        _accumulators: &[Self::Item],
+        _rng: &mut R
+    ) -> Result<bool, PCError>
+    {
+        Ok(true)
+    }
+
+    fn accumulate_items(
+        _ck: &Self::AccumulatorProverKey,
+        _accumulators: Vec<Self::Item>
+    ) -> Result<(Self::Item, Self::AccumulationProof), PCError>
+    {
+        Ok(((), ()))
+    }
+
+    fn verify_accumulated_items<R: RngCore>(
+        _current_accumulator: &Self::Item,
+        _vk: &Self::AccumulatorVerifierKey,
+        _previous_accumulators: Vec<Self::Item>,
+        _proof: &Self::AccumulationProof,
+        _rng: &mut R
+    ) -> Result<bool, PCError>
+    {
+        Ok(true)
+    }
+}
+
+// Test PCDVk
+pub struct TestPCDVk {}
+
+impl AsRef<()> for TestPCDVk
+{
+    fn as_ref(&self) -> &() { &() }
+}
+
+/// For test purpose only: The previous PCD is just a valid instance of FinalDarlinDeferredData.
+pub struct TestPrevPCD<G1: AffineCurve, G2: AffineCurve>(FinalDarlinDeferredData<G1, G2>);
```

Comment:
It would be useful to simulate also the construction of the new pub ins based on the prev proof pub ins